packages/hardhat-core/src/builtin-tasks/compile.ts

```diff
@@ -161,9 +161,22 @@ subtask(TASK_COMPILE_SOLIDITY_READ_FILE)
   .addParam("absolutePath", undefined, undefined, types.string)
   .setAction(
     async ({ absolutePath }: { absolutePath: string }): Promise<string> => {
-      return fsExtra.readFile(absolutePath, {
-        encoding: "utf8",
-      });
+      try {
+        return await fsExtra.readFile(absolutePath, {
+          encoding: "utf8",
+        });
+      } catch (e) {
+        if (!fsExtra.existsSync(absolutePath)) {
```

Comment:
I'd put these checks before the try-catch block so that it follows the error-first approach, making it a bit more readable. It's a bit of a personal preference, so if you think your way is better, feel free to ignore this comment.

Reply:
I agree that that would be more readable but I think the current approach is better performance-wise. I don't have numbers to prove it, but adding an extra file system operation for each solidity file in the project is almost surely wasteful.